client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java

```diff
@@ -2002,6 +2014,164 @@ protected static boolean convertExistsResponse(Response response) {
         return response.getStatusLine().getStatusCode() == 200;
     }
 
+    private Cancellable performClientRequestAsync(Request request, ResponseListener listener) {
+
+        ListenableFuture<Optional<String>> versionCheck = getVersionValidationFuture();
+
+        CompletableFuture<Void> cancelFlag = new CompletableFuture<Void>();
+        Cancellable result = new Cancellable() {
+            @Override
+            public void cancel() {
+                cancelFlag.complete(null);
+            }
+
+            @Override
+            void runIfNotCancelled(Runnable runnable) {
+                if (cancelFlag.isDone()) {
+                    throw newCancellationException();
+                }
+                runnable.run();
+            }
+        };
+
+        versionCheck.addListener(new ActionListener<Optional<String>>() {
+            @Override
+            public void onResponse(Optional<String> validation) {
+                if (validation.isEmpty()) {
+                    Cancellable call = client.performRequestAsync(request, listener);
+                    // Propagate cancellation
+                    cancelFlag.whenComplete((r, t) -> call.cancel());
```

Comment:
Why do we cancel the call when the future is complete? Won't the future be completed successfully but the original call still be cancelled?

Reply:
Not sure I understnad the question. The `cancelFlag` future is completed successfully when the method's result is cancelled. If/when that happens, we also need to cancel the asynchronous API call request.